README.md
# 3D Model Viewer for Flutter

<https://pub.dev/packages/model_viewer> - Fork from a package that is discontinued,
and add the web support for it.

This is a [Flutter](https://flutter.dev) widget for rendering interactive
3D models in the [glTF](https://www.khronos.org/gltf/) and
[GLB](https://wiki.fileformat.com/3d/glb/) formats.

The widget embeds Google's [`<model-viewer>`](https://modelviewer.dev)
web component in a [WebView](https://pub.dev/packages/webview_flutter).

## Screenshot

<img alt="Screenshot of astronaut model" src="./example/flutter_01.png" width="480"/>
<img alt="Screenshot of astronaut model" src="./example/flutter_02.png" width="480"/>

## Compatibility

Android, iOS, Web, with [a recent system browser version](https://modelviewer.dev/#section-browser-support).

## Notes

We use the [Google APP](https://play.google.com/store/apps/details?id=com.google.android.googlequicksearchbox), `com.google.android.googlequicksearchbox` to display interactive 3D models on Android. The model displays in 'ar_preferred' mode by default, Scene Viewer launches in AR native mode as the entry mode. If [Google Play Services for AR (ARCore, `com.google.ar.core`)](https://play.google.com/store/apps/details?id=com.google.ar.core) isn't present, Scene Viewer gracefully falls back to 3D mode as the entry mode.

## Installation

### `pubspec.yaml`

```yaml
dependencies:
   model_viewer_plus: ^(newest from https://pub.dev/packages/model_viewer_plus)
```

### `AndroidManifest.xml` (Android 9+ only)

To use this widget on Android 9+ devices, your app must be permitted to make
an HTTP connection to `http://localhost:XXXXX`. Android 9 (API level 28)
changed the default for [`android:usesCleartextTraffic`] from `true` to
`false`, so you will need to configure your app's
`android/app/src/main/AndroidManifest.xml` as follows:

```diff
--- a/example/android/app/src/main/AndroidManifest.xml
+++ b/example/android/app/src/main/AndroidManifest.xml
@@ -8,7 +8,8 @@
     <application
         android:name="io.flutter.app.FlutterApplication"
         android:label="model_viewer_example"
-        android:icon="@mipmap/ic_launcher">
+        android:icon="@mipmap/ic_launcher"
+        android:usesCleartextTraffic="true">
         <activity
             android:name=".MainActivity"
             android:launchMode="singleTop"
```

This does not affect Android 8 and earlier. See [#7] for more information.

### `app/build.gradle` (Android only)

Change minSdkVersion to 19.

    defaultConfig {
        // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
        applicationId "com.example.lab_rat"
        minSdkVersion 19
        targetSdkVersion flutter.targetSdkVersion
        versionCode flutterVersionCode.toInteger()
        versionName flutterVersionName
    }

### `Info.plist` (iOS only)

To use this widget on iOS, you need to opt-in to the embedded views preview
by adding a boolean property to your app's `ios/Runner/Info.plist` file, with
the key `io.flutter.embedded_views_preview` and the value `YES`:

```xml
<key>io.flutter.embedded_views_preview</key>
<true/>
```

### `web/index.html` (Web only)

Modify the `<head>` tag of your `web/index.html` to load the JavaScript, like so:

```html
<head>

  <!-- Other stuff -->

  <script type="module" src="https://unpkg.com/@google/model-viewer/dist/model-viewer.min.js" defer></script>
</head>
```

The [official site](https://modelviewer.dev) uses unpkg. You may replace the
value of `src` attribute with another CDN mirror's URL. Or if you are willing
to use the default js file which is included in this package's asset, you may
replace the value  with `./assets/packages/model_viewer_plus/assets/model-viewer.min.js`.

## Features

- Renders glTF and GLB models. (Also, [USDZ] models on iOS 12+.)

- Supports animated models, with a configurable auto-play setting.

- Optionally supports launching the model into an [AR] viewer.

- Optionally auto-rotates the model, with a configurable delay.

- Supports a configurable background color for the widget.

[USDZ]: https://graphics.pixar.com/usd/docs/Usdz-File-Format-Specification.html
[AR]:   https://en.wikipedia.org/wiki/Augmented_reality

## Examples

### Importing the library

```dart
import 'package:xetia_model_viewer/model_viewer.dart';
```

### Creating a `ModelViewer` widget

```dart
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: Text("Model Viewer")),
        body: ModelViewer(
          src: 'https://modelviewer.dev/shared-assets/models/Astronaut.glb',
          alt: "A 3D model of an astronaut",
          ar: true,
          autoRotate: true,
          cameraControls: true,
        ),
      ),
    );
  }
}
```

### Loading a bundled Flutter asset

```dart
class MyApp extends StatelessWidget {
// ...
          src: 'assets/MyModel.glb',
// ...
}
```

### Loading a model from the file system

This is not avaliable on Web.

```dart
class MyApp extends StatelessWidget {
// ...
          src: 'file:///path/to/MyModel.glb',
// ...
}
```

### Loading a model from the web

```dart
class MyApp extends StatelessWidget {
// ...
          src: 'https://modelviewer.dev/shared-assets/models/Astronaut.glb',
// ...
}
```

Note that due to browsers' [CORS] security restrictions, the model file
*must* be served with a `Access-Control-Allow-Origin: *` HTTP header.

## Frequently Asked Questions

### Q: Why do I get the error `net::ERR_CLEARTEXT_NOT_PERMITTED`?

You didn't configure your `AndroidManifest.xml` as per the installation
instructions earlier in this document. See also [#7].

### Q: Why does the example app just show a blank screen?

**A:** Most likely, the platform browser version on your device or emulator is
too old and does not support the features that Model Viewer needs.

For example, the stock Chrome version on the Android 10 emulator is too old
and will display a blank screen; it must be upgraded from the Play Store in
order to use this package. (The stock Chrome version on the Android 11
emulator works fine, however.) See [google/model-viewer#1109].

### Q: Why doesn't my 3D model load and/or render?

**A:** There are several reasons why your model URL could fail to load and
render:

1. It might not be possible to load the model URL due to [CORS] security
   restrictions. The server hosting the model file *must* send appropriate
   CORS response headers for Model Viewer to be able to load the file.
   See [google/model-viewer#1015].

2. It might not be possible to parse the provided glTF or GLB file.
   Some tools can produce invalid files when exporting glTF. Always
   run your model files through the [glTF Validator] to check for this.

3. The platform browser might not support the features that Model Viewer
   needs. See [google/model-viewer#1109].

[#7]:                       https://github.com/drydart/model_viewer.dart/issues/7
[CORS]:                     https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
[glTF Validator]:           https://github.khronos.org/glTF-Validator/
[google/model-viewer#1015]: https://github.com/google/model-viewer/issues/1015
[google/model-viewer#1109]: https://github.com/google/model-viewer/issues/1109
[`android:usesCleartextTraffic`]: https://developer.android.com/guide/topics/manifest/application-element#usesCleartextTraffic


CHANGELOG.md
## 1.3.0

 - **FIX**: added minRenderScale.
 - **FIX**: web onLoading.
 - **FIX**: webview background color.
 - **FIX**: poster color.
 - **FIX**: element id.
 - **FIX**: js.
 - **FIX**: function name.
 - **FIX**: element view.
 - **FIX**: variable name.
 - **FIX**: js function name.
 - **FIX**: unique js variable name.
 - **FEAT**: replace default progressbar when onLoading is handled.
 - **FEAT**: added minRenderScale.
 - **FEAT**: support change variant.
 - **FEAT**: added onLoading progress.
 - **FEAT**: using transparent when posterColor.alpha is zero.
 - **FEAT**: added --poster-color css.
 - **FEAT**: update js function name.
 - **FEAT**: improve model viewer controller.
 - **FEAT**: fix web script.
 - **FEAT**: added ModelViewerController.

# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.2.0]

### Added

- More examples in `example/lib`

### Changed

- `lib\html_builder.dart`, `lib\model_viewer_plus_mobile.dart`, `lib\model_viewer_plus_web.dart`, `lib\model_viewer_plus.dart`: implement all the attributes of `<model-viewer>` v1.11.1.
- ModelViewer.arSacle from `final String? arScale` to `final ArScale? arScale`, which may be a breaking change.
- Update `assets/model-viewer.min.js` to v1.11.1
- Update example dir's gradle version to 7.0.2

## [1.1.5] - 2022-03-15

### Changed

- `lib/src/model_viewer_plus_mobile.dart`
  - Fix [#11](https://github.com/omchiii/model_viewer_plus.dart/issues/11), add `gestureRecognizers`
  - A less elegant solution of [#8](https://github.com/omchiii/model_viewer_plus.dart/issues/8), open usdz file by [url_launcher](https://pub.dev/packages/url_launcher) in SFSafariViewController.

## [1.1.4] - 2022-03-14

### Changed

- `/lib/src/model_viewer_plus_mobile.dart`, update according to the [newest document](https://developers.google.com/ar/develop/scene-viewer#3d-or-ar). Fix [#9](https://github.com/omchiii/model_viewer_plus.dart/issues/9).
  - Insted of `com.google.ar.core`, now we use `com.google.android.googlequicksearchbox`. This should support the widest possible range of devices.
  - Mode defaults to `ar_preferred`. Scene Viewer launches in AR native mode as the entry mode. If Google Play Services for AR isn't present, Scene Viewer gracefully falls back to 3D mode as the entry mode.
- Add `arModes` to example, closer to [modelviewer.dev](https://modelviewer.dev)'s offical example.
- Update `example\android\app\build.gradle` `compileSdkVersion` to 31
- Update `android_intent_plus` to `3.1.1`
- Update `webview_flutter` to `3.0.1`

### Removed

- `/lib/src/http_proxy.dart`: empty file

## [1.1.3] - 2022-03-14

### Changed

- `ModelViewer`'s default `backgroundColor` from `Colors.white` to `Colors.transparent`, due to [#12](https://github.com/omchiii/model_viewer_plus.dart/issues/12)
- `proxy`'s null check fix and `setState() {}` for it, due to [#10](https://github.com/omchiii/model_viewer_plus.dart/issues/10)

## [1.1.2] - 2022-02-17

### Added

- `/lib/src/shim/` with `dart_html_fake.dart` and `dart_ui_fake.dart`. Fixing `ERROR: The name platformViewRegistry' is being referenced through the prefix 'ui', but it isn't defined in any of the libraries imported using that prefix.` and `INFO: Avoid using web-only libraries outside Flutter web plugin` to improve the [score on pub.dev](https://pub.dev/packages/model_viewer_plus/score).

### Changed

- example's `/etc/assets` -> `/assets`

## [1.1.1] - 2022-02-17

### Added

- `/assets/model-viewer.min.js` (v1.10.1, which is actually identical to v1.10.0)

### Changed

- `/etc/assets` -> `/assets`
- `example/flutter_02.png`: Updated the Screenshot.
- `README.md`: README Update.
- `lib/src/model_viewer_plus_web.dart`: Due to the change of `model-viewer.js` -> `model-viewer.min.js`
- `lib/src/model_viewer_plus_mobile.dart`: Due to the change of `model-viewer.js` -> `model-viewer.min.js`, added CircularProgressIndicator while mobile platform loading

### Removed

- `/etc/assets/model-viewer.js`: To slim the package size.

## [1.1.0] - 2022-02-15

### Added

- Web Support
  - `lib/src/model_viewer_plus_stub.dart`
  - `lib/src/model_viewer_plus_mobile.dart`
  - `lib/src/model_viewer_plus_web.dart`

### Changed

- `lib/model_viewer_plus.dart`
- `lib/src/model_viewer_plus.dart`


lib/src/model_viewer_plus_mobile.dart
/* This is free and unencumbered software released into the public domain. */

import 'dart:async' show Completer;
import 'dart:convert' show json, utf8;
import 'dart:io'
    show File, HttpRequest, HttpServer, HttpStatus, InternetAddress, Platform;
import 'dart:math' as math;
import 'dart:typed_data' show Uint8List;

import 'package:android_intent_plus/android_intent.dart' as android_content;
import 'package:flutter/foundation.dart';
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart' show rootBundle;
import 'package:xetia_model_viewer/src/model_viewer_controller.dart';
import 'package:path/path.dart' as p;
import 'package:url_launcher/url_launcher_string.dart';
import 'package:webview_flutter/webview_flutter.dart';

import 'html_builder.dart';
import 'model_viewer_plus.dart';

final _random = math.Random();

class ModelViewerState extends State<ModelViewer> {
  final Completer<WebViewController> _controller =
      Completer<WebViewController>();

  HttpServer? _proxy;
  late String _proxyURL;

  @override
  void initState() {
    super.initState();
    _initProxy();
  }

  @override
  void dispose() {
    super.dispose();
    if (_proxy != null) {
      _proxy!.close(force: true);
      _proxy = null;
    }
  }

  @override
  void didUpdateWidget(final ModelViewer oldWidget) {
    super.didUpdateWidget(oldWidget);
    // TODO
  }

  @override
  Widget build(final BuildContext context) {
    if (_proxy == null) {
      return Center(
        child: CircularProgressIndicator(
          semanticsLabel: 'Loading Model Viewer...',
        ),
      );
    } else {
      return WebView(
        backgroundColor: widget.backgroundColor,
        initialUrl: null,
        javascriptMode: JavascriptMode.unrestricted,
        initialMediaPlaybackPolicy: AutoMediaPlaybackPolicy.always_allow,
        gestureRecognizers: <Factory<OneSequenceGestureRecognizer>>{
          Factory<OneSequenceGestureRecognizer>(
            () => EagerGestureRecognizer(),
          ),
        },
        onWebViewCreated: (final WebViewController webViewController) async {
          print('>>>> ModelViewer initializing... <$_proxyURL>'); // DEBUG
          await webViewController.loadUrl(_proxyURL);
          _controller.complete(webViewController);
        },
        javascriptChannels: {
          JavascriptChannel(
            name: 'OnLoadedEvent',
            onMessageReceived: (s) async {
              final webViewController = await _controller.future;
              if (widget.minRenderScale != null) {
                await webViewController.runJavascript(
                    'setMinimumRenderScale(${widget.minRenderScale})');
              }

              final getMaterialsResult = await webViewController
                  .runJavascriptReturningResult('getMaterials()');
              // on Android it will be returned as a string of string
              // on iOS it will be returned as a string of JSON

              final materialJson = json.decode(getMaterialsResult);
              var materials = <String>[];

              if (materialJson is String) {
                final res = json.decode(materialJson);
                if (res is List) {
                  materials = res.cast<String>().toSet().toList();
                }
              } else if (materialJson is List) {
                materials = materialJson.cast<String>().toSet().toList();
              }

              final getVariantsResult = await webViewController
                  .runJavascriptReturningResult('getVariants()');

              final variantsJson = json.decode(getVariantsResult);
              var variants = <String>[];

              if (variantsJson is String) {
                final res = json.decode(variantsJson);
                if (res is List) {
                  variants = res.cast<String>().toSet().toList();
                }
              } else if (variantsJson is List) {
                variants = variantsJson.cast<String>().toSet().toList();
              }

              final controller = _ModelViewerControllerImpl(
                materials: materials,
                variants: variants,
                webViewController: webViewController,
              );
              widget.onCreated?.call(controller);
            },
          ),
          JavascriptChannel(
            name: 'OnProgressEvent',
            onMessageReceived: (m) async {
              final value = double.tryParse(m.message);
              widget.onLoading?.call(value ?? 0);
            },
          ),
        },
        navigationDelegate: (final NavigationRequest navigation) async {
          print('>>>> ModelViewer wants to load: <${navigation.url}>'); // DEBUG
          if (!Platform.isAndroid) {
            if (Platform.isIOS && navigation.url == widget.iosSrc) {
              await launchUrlString(
                navigation.url,
              );
              return NavigationDecision.prevent;
            }
            return NavigationDecision.navigate;
          }
          if (!navigation.url.startsWith("intent://")) {
            return NavigationDecision.navigate;
          }
          try {
            // Original, just keep as a backup
            // See: https://developers.google.com/ar/develop/java/scene-viewer
            // final intent = android_content.AndroidIntent(
            //   action: "android.intent.action.VIEW", // Intent.ACTION_VIEW
            //   data: "https://arvr.google.com/scene-viewer/1.0",
            //   arguments: <String, dynamic>{
            //     'file': widget.src,
            //     'mode': 'ar_preferred',
            //   },
            //   package: "com.google.ar.core",
            //   flags: <int>[
            //     Flag.FLAG_ACTIVITY_NEW_TASK
            //   ], // Intent.FLAG_ACTIVITY_NEW_TASK,
            // );

            // 2022-03-14 update
            final String fileURL;
            if (['http', 'https'].contains(Uri.parse(widget.src).scheme)) {
              fileURL = widget.src;
            } else {
              fileURL = p.joinAll([_proxyURL, 'model']);
            }
            final intent = android_content.AndroidIntent(
              action: "android.intent.action.VIEW", // Intent.ACTION_VIEW
              // See https://developers.google.com/ar/develop/scene-viewer#3d-or-ar
              // data should be something like "https://arvr.google.com/scene-viewer/1.0?file=https://raw.githubusercontent.com/KhronosGroup/glTF-Sample-Models/master/2.0/Avocado/glTF/Avocado.gltf"
              data: Uri(
                  scheme: 'https',
                  host: 'arvr.google.com',
                  path: '/scene-viewer/1.0',
                  queryParameters: {
                    // 'title': '', // TODO: maybe set by the user
                    // TODO: further test, and make it 'ar_preferred'
                    'mode': 'ar_preferred',
                    'file': fileURL,
                  }).toString(),
              // package changed to com.google.android.googlequicksearchbox
              // to support the widest possible range of devices
              package: "com.google.android.googlequicksearchbox",
              arguments: <String, dynamic>{
                'browser_fallback_url':
                    'market://details?id=com.google.android.googlequicksearchbox'
              },
            );
            await intent.launch().onError((error, stackTrace) {
              print('>>>> ModelViewer Intent Error: $error'); // DEBUG
            });
          } catch (error) {
            print('>>>> ModelViewer failed to launch AR: $error'); // DEBUG
          }
          return NavigationDecision.prevent;
        },
        onPageStarted: (final String url) {
          //print('>>>> ModelViewer began loading: <$url>'); // DEBUG
        },
        onPageFinished: (final String url) {
          //print('>>>> ModelViewer finished loading: <$url>'); // DEBUG
        },
        onWebResourceError: (final WebResourceError error) {
          print(
              '>>>> ModelViewer failed to load: ${error.description} (${error.errorType} ${error.errorCode})'); // DEBUG
        },
      );
    }
  }

  String _buildHTML(final String htmlTemplate) {
    return HTMLBuilder.build(
      htmlTemplate: htmlTemplate,
      src: '/model',
      alt: widget.alt,
      poster: widget.poster,
      seamlessPoster: widget.seamlessPoster,
      loading: widget.loading,
      reveal: widget.reveal,
      withCredentials: widget.withCredentials,
      // AR Attributes
      ar: widget.ar,
      arModes: widget.arModes,
      arScale: widget.arScale,
      arPlacement: widget.arPlacement,
      iosSrc: widget.iosSrc,
      xrEnvironment: widget.xrEnvironment,
      // Staing & Cameras Attributes
      cameraControls: widget.cameraControls,
      enablePan: widget.enablePan,
      touchAction: widget.touchAction,
      disableZoom: widget.disableZoom,
      orbitSensitivity: widget.orbitSensitivity,
      autoRotate: widget.autoRotate,
      autoRotateDelay: widget.autoRotateDelay,
      rotationPerSecond: widget.rotationPerSecond,
      interactionPolicy: widget.interactionPolicy,
      interactionPrompt: widget.interactionPrompt,
      interactionPromptStyle: widget.interactionPromptStyle,
      interactionPromptThreshold: widget.interactionPromptThreshold,
      cameraOrbit: widget.cameraOrbit,
      cameraTarget: widget.cameraTarget,
      fieldOfView: widget.fieldOfView,
      maxCameraOrbit: widget.maxCameraOrbit,
      minCameraOrbit: widget.minCameraOrbit,
      maxFieldOfView: widget.maxFieldOfView,
      minFieldOfView: widget.minFieldOfView,
      bounds: widget.bounds,
      interpolationDecay: widget.interpolationDecay,
      // Lighting & Env Attributes
      skyboxImage: widget.skyboxImage,
      environmentImage: widget.environmentImage,
      exposure: widget.exposure,
      shadowIntensity: widget.shadowIntensity,
      shadowSoftness: widget.shadowSoftness,
      // Animation Attributes
      animationName: widget.animationName,
      animationCrossfadeDuration: widget.animationCrossfadeDuration,
      autoPlay: widget.autoPlay,
      // Scene Graph Attributes
      variantName: widget.variantName,
      orientation: widget.orientation,
      scale: widget.scale,

      // CSS Styles
      backgroundColor: widget.backgroundColor,
      // Loading CSS
      posterColor: widget.posterColor,
      // Annotations CSS
      minHotspotOpacity: widget.minHotspotOpacity,
      maxHotspotOpacity: widget.maxHotspotOpacity,

      // Others
      innerModelViewerHtml: _innerModelViewerHtml(),
      relatedCss: _relatedCss(),
      relatedJs: _relatedJs(),
      id: widget.id,
    );
  }

  String? _relatedCss() {
    if (widget.onLoading == null) {
      return widget.relatedCss;
    }

    return '''
.progress-bar {
  display: none;
  visibility: hidden;
}
${widget.innerModelViewerHtml ?? ''}''';
  }

  String? _innerModelViewerHtml() {
    if (widget.onLoading == null) {
      return widget.innerModelViewerHtml;
    }

    return '<div class="progress-bar" slot="progress-bar"></div>${widget.innerModelViewerHtml ?? ''}';
  }

  String _createVariableName() {
    return 'modelViewer${_random.nextInt(10000000)}';
  }

  String? _variableName;
  String get variableName {
    return _variableName ??= _createVariableName();
  }

  String _relatedJs() {
    return '''
  const $variableName = document.querySelector("model-viewer#${widget.id}");
  $variableName.addEventListener('load', function () {
      OnLoadedEvent.postMessage(true);
  });

  $variableName.addEventListener('progress', function (event) {
      OnProgressEvent.postMessage(event.detail.totalProgress);
  });

  function updateMaterialColor(name, color) {
    const material = $variableName.model.getMaterialByName(name);
    if (material){
      material.pbrMetallicRoughness.setBaseColorFactor(color);
    }
  }

  function getMaterials() {
    const materials = $variableName.model.materials;
    const result = JSON.stringify(materials.map(material => material.name));
    return result;
  }

  function getVariants() {
    const names = $variableName.availableVariants;
    const result = JSON.stringify(names);
    return result;
  }

  function setVariant(variantName) {
    $variableName.variantName = variantName;
  }

  function setMinimumRenderScale(value){
    $variableName.minimumRenderScale = value;
  }

  ${widget.relatedJs}
  ''';
  }

  Future<void> _initProxy() async {
    final url = Uri.parse(widget.src);
    _proxy = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);

    setState(() {
      _proxy;
      final host = _proxy!.address.address;
      final port = _proxy!.port;
      _proxyURL = "http://$host:$port/";
    });

    _proxy!.listen((final HttpRequest request) async {
      //print("${request.method} ${request.uri}"); // DEBUG
      //print(request.headers); // DEBUG
      final response = request.response;

      switch (request.uri.path) {
        case '/':
        case '/index.html':
          final htmlTemplate = await rootBundle
              .loadString('packages/model_viewer_plus/assets/template.html');
          final html = utf8.encode(_buildHTML(htmlTemplate));
          response
            ..statusCode = HttpStatus.ok
            ..headers.add("Content-Type", "text/html;charset=UTF-8")
            ..headers.add("Content-Length", html.length.toString())
            ..add(html);
          await response.close();
          break;

        case '/model-viewer.min.js':
          final code = await _readAsset(
              'packages/model_viewer_plus/assets/model-viewer.min.js');
          response
            ..statusCode = HttpStatus.ok
            ..headers
                .add("Content-Type", "application/javascript;charset=UTF-8")
            ..headers.add("Content-Length", code.lengthInBytes.toString())
            ..add(code);
          await response.close();
          break;

        case '/model':
          if (url.isAbsolute && !url.isScheme("file")) {
            // debugPrint(url.toString());
            await response.redirect(url); // TODO: proxy the resource
          } else {
            final data = await (url.isScheme("file")
                ? _readFile(url.path)
                : _readAsset(url.path));
            response
              ..statusCode = HttpStatus.ok
              ..headers.add("Content-Type", "application/octet-stream")
              ..headers.add("Content-Length", data.lengthInBytes.toString())
              ..headers.add("Access-Control-Allow-Origin", "*")
              ..add(data);
            await response.close();
          }
          break;

        case '/favicon.ico':
          final text = utf8.encode("Resource '${request.uri}' not found");
          response
            ..statusCode = HttpStatus.notFound
            ..headers.add("Content-Type", "text/plain;charset=UTF-8")
            ..headers.add("Content-Length", text.length.toString())
            ..add(text);
          await response.close();
          break;

        default:
          if (request.uri.isAbsolute) {
            await response.redirect(request.uri);
          } else if (request.uri.hasAbsolutePath) {
            // Some gltf models need other resources from the origin
            var pathSegments = [...url.pathSegments];
            pathSegments.removeLast();
            var tryDestination = p.joinAll([
              url.origin,
              ...pathSegments,
              request.uri.path.replaceFirst('/', '')
            ]);
            debugPrint("Try: ${tryDestination}");
            await response.redirect(Uri.parse(tryDestination));
          } else {
            debugPrint('404 with ${request.uri}');
            final text = utf8.encode("Resource '${request.uri}' not found");
            response
              ..statusCode = HttpStatus.notFound
              ..headers.add("Content-Type", "text/plain;charset=UTF-8")
              ..headers.add("Content-Length", text.length.toString())
              ..add(text);
            await response.close();
            break;
          }
      }
    });
  }

  Future<Uint8List> _readAsset(final String key) async {
    final data = await rootBundle.load(key);
    return data.buffer.asUint8List(data.offsetInBytes, data.lengthInBytes);
  }

  Future<Uint8List> _readFile(final String path) async {
    return await File(path).readAsBytes();
  }
}

class _ModelViewerControllerImpl implements ModelViewerController {
  final WebViewController webViewController;

  @override
  final List<String> materials;
  @override
  final List<String> variants;

  _ModelViewerControllerImpl({
    required this.webViewController,
    required this.materials,
    required this.variants,
  });

  @override
  Future<void> changeColor(String materialName, Color color) async {
    final colorString = [
      color.red / 256,
      color.green / 256,
      color.blue / 256,
      color.alpha / 256,
    ];
    await webViewController.runJavascriptReturningResult(
        'updateMaterialColor("$materialName", $colorString)');
  }

  @override
  Future<void> setVariant(String? variant) async {
    await webViewController
        .runJavascriptReturningResult('setVariant("$variant")');
  }
}
